hera_cal/smooth_cal.py

```diff
@@ -105,9 +106,109 @@ def time_filter(gains, wgts, times, filter_scale=1800.0, nMirrors=0):
+                        tol=1e-09, filter_mode='rect', maxiter=100, window='tukey', **win_kwargs):
+    '''Filter calibration solutions in both time and frequency simultaneously. First rephases to remove
+    a time-smoothed delay from the gains, then performs the low-pass 2D filter in time and frequency,
+    then puts back in the delay rephasor. Uses aipy.deconv.clean to account for weights/flags.
+
+    Arguments:
+        gains: ndarray of shape=(Ntimes,Nfreqs) of complex calibration solutions to filter
+        wgts: ndarray of shape=(Ntimes,Nfreqs) of real linear multiplicative weights
+        freqs: ndarray of frequency channels in Hz
+        times: ndarray of shape=(Ntimes) of Julian dates as floats in units of days
+        freq_scale: frequency scale in MHz to use for the low-pass filter. freq_scale^-1 corresponds
+            to the half-width (i.e. the width of the positive part) of the region in fourier
+            space, symmetric about 0, that is filtered out.
+        time_scale: time scale in seconds. Defined analogously to freq_scale.
+        tol: CLEAN algorithm convergence tolerance (see aipy.deconv.clean)
+        filter_mode: either 'rect' or 'plus':
+            'rect': perform 2D low-pass filter, keeping modes in a small rectangle around delay = 0 
+                    and fringe rate = 0
+            'plus': produce a separable calibration solution by only keeping modes with 0 delay,
+                    0 fringe rate, or both
+        window: window function for filtering applied to the filtered axis.
+            See aipy.dsp.gen_window for options.
+        maxiter: Maximum number of iterations for aipy.deconv.clean to converge.
+        win_kwargs : any keyword arguments for the window function selection in aipy.dsp.gen_window.
+            Currently, the only window that takes a kwarg is the tukey window with a alpha=0.5 default.
+
+    Returns:
+        filtered: filtered gains, ndarray of shape=(Ntimes,Nfreqs)
+        info: dictionary of metadata from aipy.deconv.clean
+    '''
+    df = np.median(np.diff(freqs))
+    dt = np.median(np.diff(times)) * 24.0 * 3600.0  # in seconds
+    delays = np.fft.fftfreq(freqs.size, df)
+    fringes = np.fft.fftfreq(times.size, dt)
+    delay_scale = (freq_scale * 1e6)**-1  # Puts it in seconds
+    fringe_scale = (time_scale)**-1  # in Hz
+
+    # find per-integration delays, smooth on the time_scale of gain smoothing, and rephase 
+    taus = fft_dly(gains, df, wgts, medfilt=False, solve_phase=False)[0].astype(np.complex)  # delays are in seconds
+    if not np.all(taus == 0):  # this breaks CLEAN, but it means we don't need smoothing anyway
+        taus = uvtools.dspec.high_pass_fourier_filter(taus.T, np.sum(wgts, axis=1, keepdims=True).T,
+                                                      fringe_scale, dt, tol=tol, maxiter=maxiter)[0].T
+    rephasor = np.exp(-2.0j * np.pi * np.outer(np.abs(taus), freqs))
```

Comment:
Should rephasing be an option, rather than being done automatically? I _think_ the idea is that if delays are off, this will center things so you can smooth around the most significant delay. But if, for example, you're calculating gain corrections, where you don't expect to have significant peaks at center delay, this might do weird things.

Reply:
Sorry, I don't follow. What do you mean by gain corrections? 

If what you're smoothing is basically flat, this will find the delay to be 0 and it will smooth around 0 delay. If the main delay is not 0, we need to smooth around that otherwise you end up getting destructive interference and introducing very strange spectral structure.

Reply:
I don't know if I understand the bit about destructive interference, but I also don't understand my original question. I think I was operating with less than enough sleep.

Reply:
haha, ok.